go/cmd/graphql/main.go

```diff
@@ -107,14 +115,26 @@ func setupAPI(
 	getFunc func(http.ResponseWriter, *http.Request),
 	deleteFunc func(http.ResponseWriter, *http.Request),
 ) {
-	router.HandleFunc("", listFunc).Methods("GET")
-	router.HandleFunc("/", listFunc).Methods("GET")
-	router.HandleFunc("", writeFunc).Methods("POST")
-	router.HandleFunc("/", writeFunc).Methods("POST")
-	router.HandleFunc("/{name}", getFunc).Methods("GET")
-	router.HandleFunc("/{name}/", getFunc).Methods("GET")
-	router.HandleFunc("/{name}", deleteFunc).Methods("DELETE")
-	router.HandleFunc("/{name}/", deleteFunc).Methods("DELETE")
+	// Ensure that each call is authenticated before proceeding
+	router.HandleFunc("", authenticateTenantThenCall(listFunc)).Methods("GET")
+	router.HandleFunc("/", authenticateTenantThenCall(listFunc)).Methods("GET")
+	router.HandleFunc("", authenticateTenantThenCall(writeFunc)).Methods("POST")
+	router.HandleFunc("/", authenticateTenantThenCall(writeFunc)).Methods("POST")
+	router.HandleFunc("/{name}", authenticateTenantThenCall(getFunc)).Methods("GET")
+	router.HandleFunc("/{name}/", authenticateTenantThenCall(getFunc)).Methods("GET")
+	router.HandleFunc("/{name}", authenticateTenantThenCall(deleteFunc)).Methods("DELETE")
+	router.HandleFunc("/{name}/", authenticateTenantThenCall(deleteFunc)).Methods("DELETE")
+}
+
+// Wraps `f` in a preceding check that authenticates the request headers for the expected tenant name.
+// The check is skipped if `disableAPIAuthentication` is true.
+func authenticateTenantThenCall(f func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !disableAPIAuthentication && !middleware.DataAPIRequestAuthenticator(w, r, tenantName) {
```

Comment:
Happy that I did refactor this a bit before you started your work, happy that you could simply adopt this.

I'm suuuuper new to Golang -- any feedback about the mess I created is welcome @nickbp